lib/wasi/src/syscalls/mod.rs

```diff
@@ -297,75 +325,203 @@ where
-            // Optional timeout
-            _ = timeout => Err(Errno::Timedout),
-        })
-    };
+    let tasks = env.tasks().clone();
+    let poller = Poller { ctx, pinned_work };
+    block_on_with_timeout(&tasks, timeout, poller)
+}
 
-    // Fast path
-    if nonblocking {
-        let waker = WasiDummyWaker.into_waker();
-        let mut cx = Context::from_waker(&waker);
-        let _guard = tasks.runtime_enter();
-        let mut pinned_work = Box::pin(work);
-        if let Poll::Ready(res) = pinned_work.as_mut().poll(&mut cx) {
-            return res;
+/// Future that will be polled by asyncify methods
+pub trait AsyncifyFuture {
+    type Output;
+
+    fn poll(
+        &mut self,
+        env: &WasiEnv,
+        store: &dyn AsStoreRef,
+        cx: &mut Context<'_>,
+    ) -> Poll<Self::Output>;
+}
+
+// This poller will process any signals when the main working function is idle
+struct AsyncifyPoller<'a, 'b, 'c, Fut, T>
+where
+    Fut: AsyncifyFuture<Output = Result<T, Errno>>,
+{
+    env: &'a WasiEnv,
+    store: &'b dyn AsStoreRef,
+    work: &'c mut Fut,
+}
+impl<'a, 'b, 'c, Fut, T> Future for AsyncifyPoller<'a, 'b, 'c, Fut, T>
+where
+    Fut: AsyncifyFuture<Output = Result<T, Errno>>,
+{
+    type Output = Result<Fut::Output, WasiError>;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let env = self.env;
+        let store = self.store;
+        if let Poll::Ready(res) = self.work.poll(env, store, cx) {
+            return Poll::Ready(Ok(res));
         }
-        return Ok(Err(Errno::Again));
+        if let Some(exit_code) = self.env.should_exit() {
+            return Poll::Ready(Err(WasiError::Exit(exit_code)));
+        }
+        if let Some(signals) = self.env.thread.pop_signals_or_subscribe(cx.waker()) {
+            return Poll::Ready(Ok(Err(Errno::Intr)));
+        }
+        Poll::Pending
     }
+}
 
-    // Slow path, block on the work and process process
-    tasks.block_on(work)
+pub enum AsyncifyAction<'a> {
+    /// Indicates that asyncify callback finished and the
+    /// caller now has ownership of the ctx again
+    Finish(FunctionEnvMut<'a, WasiEnv>),
+    /// Indicates that asyncify should unwind by immediately exiting
+    /// the current function
+    Unwind,
 }
 
 /// Asyncify takes the current thread and blocks on the async runtime associated with it
 /// thus allowed for asynchronous operations to execute. It has built in functionality
 /// to (optionally) timeout the IO, force exit the process, callback signals and pump
 /// synchronous IO engine
-pub(crate) fn __asyncify_light<T, Fut>(
-    env: &WasiEnv,
+///
+/// This will either return the `ctx` as the asyncify has completed successfully
+/// or it will return an WasiError which will exit the WASM call using asyncify
+/// and instead process it on a shared task
+///
+pub(crate) fn __asyncify_with_deep_sleep<M: MemorySize, Fut>(
```

Comment:
This behaviour needs to be opt-in and specified in `WasiEnvBuilder`. 

Good place to put th esetting might be in `ControlPlane`?

Reply:
I have added an opt-out in `ControlPlane`